symengine/tests/basic/test_parser.cpp

```diff
@@ -0,0 +1,258 @@
+    REQUIRE(eq(*res, *neg(div(pi, integer(2)))));
+
+    s = "asin(sin(x))";
+    res = p.parse(s);
+    REQUIRE(eq(*res, *asin(sin(x))));
+
+    s = "beta(x,y)";
+    res = p.parse(s);
+    REQUIRE(eq(*res, *beta(x,y)));
+
+    s = "beta(sin(x+3), gamma(2^y+sin(y)))";
+    res = p.parse(s);
+    REQUIRE(eq(*res, *beta(sin(add(x, integer(3))), gamma(add(sin(y), pow(integer(2), y))))));
+
+    s = "y^(abs(sin(3) + x)) + sinh(2)";
+    res = p.parse(s);
+    REQUIRE(eq(*res, *add(pow(y, abs(add(sin(integer(3)), x))), sinh(integer(2)))));
+
+    s = "2 + zeta(2, x) + zeta(ln(3))";
+    res = p.parse(s);
+    REQUIRE(eq(*res, *add(integer(2), add(zeta(integer(2), x), zeta(log(integer(3)))))));
+
+    s = "sin(asin(x)) + y";
+    res = p.parse(s);
+    REQUIRE(eq(*res, *add(x, y)));
+
+    s = "log(x, gamma(y))*sin(3)";
+    res = p.parse(s);
+    REQUIRE(eq(*res, *mul(log(x, gamma(y)), sin(integer(3)))));
+}
+
+TEST_CASE("Parsing: constants", "[parser]")
+{
+    std::string s;
+    ExpressionParser p;
+    RCP<const Basic> res;
+    RCP<const Basic> x = symbol("x");
+
+    s = "sin(pi/2)";
+    res = p.parse(s);
+    REQUIRE(eq(*res, *one));
+
+    s = "log(e)";
+    res = p.parse(s);
+    REQUIRE(eq(*res, *one));
+
+    s = "ln(e/e) + sin(pi*2/2) + 3*x";
+    res = p.parse(s);
+    REQUIRE(eq(*res, *mul(integer(3), x)));
+
+    s = "(3+4*I)/(5+cos(pi/2)*I)";
+    res = p.parse(s);
+    REQUIRE(eq(*res, *div(Complex::from_two_nums(*integer(3), *integer(4)), integer(5))));
+
+    s = "(2*I +6*I)*3*I + 4*I";
+    res = p.parse(s);
+    REQUIRE(eq(*res, *Complex::from_two_nums(*integer(-24), *integer(4))));
+}
+
+TEST_CASE("Parsing: function_symbols", "[parser]")
+{
+    std::string s;
+    ExpressionParser p;
+    RCP<const Basic> res;
+    RCP<const Basic> x = symbol("x");
+    RCP<const Basic> y = symbol("y");
+    RCP<const Basic> z = symbol("wt");
+
+    s = "f(x)";
+    res = p.parse(s);
+    REQUIRE(eq(*res, *function_symbol("f", x)));
+
+    s = "my_func(x, wt) + sin(f(y))";
+    res = p.parse(s);
+    REQUIRE(eq(*res, *add(function_symbol("my_func", {x, z}), sin(function_symbol("f", y)))));
+
+    s = "func(x, y, wt) + f(sin(x))";
+    res = p.parse(s);
+    REQUIRE(eq(*res, *add(function_symbol("func", {x, y, z}), function_symbol("f", sin(x)))));
+}
+
+TEST_CASE("Parsing: doubles", "[parser]")
+{
+    std::string s;
+    ExpressionParser p;
+    RCP<const Basic> res;
+    RCP<const Basic> x = symbol("x");
+
+    s = "1.324";
+    res = p.parse(s);
+    REQUIRE(eq(*res, *real_double(1.324)));
+
+    s = "0.0324*x + 2*3";
+    res = p.parse(s);
+    REQUIRE(eq(*res, *add(mul(real_double(0.0324), x), integer(6))));
+
+    s = "1.324/(2+3)";
+    res = p.parse(s);
+    REQUIRE(eq(*res, *real_double(0.2648)));
+
+    // can you suggest more test cases
+    // i am still not clear about the fundamentals of realdouble
+
+    s = "sqrt(2.0)+5";
+    // s = "sqrt(2)+5" this fails! (will the user add `.0` to integers to get them `computed`)
```

Comment:
You can remove this line. It's supposed to work that way. If the user just wants numerics, then they are better off using a specialized library.